Migrate `snipes` and `snipesFor` in tests to `cleanByImpersonation` or `marketOrder*`

espendk opened the pull request:
This PR completes the replacement of `snipes` with `cleanByImpersonation` started in  #437 by migrating all tests away from `snipes` and `snipesFor`. 

TODO:
- [ ]  The main test in `Scenarii.t.sol` relied on snipe. It is not clear to me whether it's a test of sniping (and thus can be deleted) or if it also tests something else (and thus should be removed)? @jkrivine maybe knows?

espendk committed:
chore: Mostly migrate TestTaker from `snipes` to `clean`

Gas tests in Gas.t.sol have been commented out as they are expected to be replaced by a new test suite.

Scenarii.t.sol has not been migrated as it's non-trivial and unclear if it makes sense w/o snipe.

espendk committed:
fix: Match names of return values for cleanByImpersonation in IMangrove

espendk committed:
WIP migrate snipe tests

espendk committed:
fix(test): Add missing vm.expectEmits for Credit events

espendk committed:
feat: Migrate `snipes` and `snipesFor` in tests to `cleanByImpersonation` or `marketOrder*`

espendk committed:
fix: Restore event order for `cleanByImpersonation`

espendk committed:
chore: Update foundry.toml with opposite of new default for `optimizer`

adhusson commented on test/core/Gatekeeping.t.sol:
```diff
@@ -305,37 +305,10 @@ contract GatekeepingTest is IMaker, MangroveTest {
     tkr.marketOrder(0, 1 << 160);
   }
 
-  //FIXME Should add similar tests that make sure volume*price is not too big.
-  function test_gives_volume_above_96bits_fails_snipes() public {
-    uint ofr = mkr.newOfferByVolume(1 ether, 1 ether, 100_000);
-    uint[4][] memory targets = wrap_dynamic([ofr, 0, 1 << 96, type(uint).max]);
-    vm.expectRevert("mgv/snipes/volume/96bits");
-    testMgv.snipesInTest($(base), $(quote), targets, true);
-  }
-
-  function test_wants_volume_above_96bits_fails_snipes() public {
-    uint ofr = mkr.newOfferByVolume(1 ether, 1 ether, 100_000);
-    uint[4][] memory targets = wrap_dynamic([ofr, 0, 1 << 96, type(uint).max]);
-    vm.expectRevert("mgv/snipes/volume/96bits");
-    testMgv.snipesInTest($(base), $(quote), targets, false);
-  }
-
```
Is there a similar test for clean's max volume?

espendk replied:
Yes, in `TakerOperations.t.sol` there's `test_gives_volume_above_96bits_fails_clean` and `test_gives_tick_outside_range_fails_clean`.

adhusson commented on test/core/Gatekeeping.t.sol:
```diff
@@ -305,37 +305,10 @@ contract GatekeepingTest is IMaker, MangroveTest {
     assertEq(mgv.allowances($(base), $(quote), address(tkr), $(this)), 0, "initial allowance should be 0");
   }
 
-  function test_cannot_snipesFor_for_without_allowance() public {
-    deal($(base), address(mkr), 1 ether);
-    mkr.approveMgv(base, 1 ether);
-    uint ofr = mkr.newOfferByVolume(1 ether, 1 ether, 100_000);
-    Tick offerTick = pair.offers(ofr).tick();
-
-    vm.expectRevert("mgv/lowAllowance");
-    testMgv.snipesForInTest(
-      $(base), $(quote), wrap_dynamic([ofr, uint(Tick.unwrap(offerTick)), 1 ether, 300_000]), true, address(tkr)
-    );
-  }
```
Is there a test that checks clean cannot decrease the `taker` balance?

espendk replied:
In  `TakerOperations.t.sol` there's `test_cleaning_by_impersonation_succeeds_and_does_not_transfer_funds`.

Do you think we need more/smth different?

adhusson commented on test/core/Gatekeeping.t.sol:
```diff
@@ -442,18 +415,18 @@ contract GatekeepingTest is IMaker, MangroveTest {
     uint other_ofr = mgv.newOfferByVolume($(quote), $(base), 1 ether, 1 ether, 100_000, 0);
 
     trade_cb = abi.encodeCall(this.updateOfferOK, ($(quote), $(base), other_ofr));
-    uint ofr = mgv.newOfferByVolume($(base), $(quote), 1 ether, 1 ether, 400_000, 0);
```
Here and in the other tests where `x_on_y_succeeds|fails` I'm not seeing a check that the callbacks were indeed called. There wasn't a check before but it was very likely since the offer with the callback was being sniped. Here if there is a change to `setUp` in this contract or the parent contracts and a cheap offer is posted to the market and succeeds, then won't the tests become vacuously true?

espendk replied:
The assertion two lines down (L448) is checking that the gasreq of the offer posted in line 442 has been changed as defined in the posthook. Both numbers are defined in the test, are unaffected by `setUp`, and (I believe) can only be changed by the posthook (the offer id is passed from the test). So it should be ok?

espendk replied:
Thinking some more on this: I think was we need is an assertion that the expected offer was executed since it's not guaranteed by `marketOrder` whereas it was with `snipe`. The tests in `Posthook.t.sol` should allow us to assume that the posthook is called if the offer is executed.

espendk replied:
I'm implementing this and it's cleaner + makes things easier to read

espendk replied:
Addressed in [6d1388d](https://github.com/mangrovedao/mangrove-core/pull/454/commits/6d1388d7700b735d4d362afbe231ab09109b62bd)

adhusson commented on test/core/MakerOperations.t.sol:
```diff
@@ -83,8 +83,8 @@ contract MakerOperationsTest is MangroveTest, IMaker {
-    require(tkr.take(ofr, 0.05 ether), "take must work or test is void");
+    mgv.newOfferByVolume($(base), $(quote), 0.05 ether, 0.05 ether, 200_000, 0);
+    require(tkr.marketOrderWithSuccess(0.05 ether), "take must work or test is void");
   }
```
Similarly in this file I this it is less clear that the callback are being called. I think something like a `callbackWasCalled()` on each test would be enough, and `callbackWasCalled` checks that some state variable has been flipped from false to true?

edit: I see we do it in `MakerPosthook.t.sol`!

espendk replied:
addressed in [6d1388d](https://github.com/mangrovedao/mangrove-core/pull/454/commits/6d1388d7700b735d4d362afbe231ab09109b62bd)

adhusson commented on test/core/MakerPosthook.t.sol:
```diff
@@ -422,8 +421,8 @@ contract MakerPosthookTest is MangroveTest, IMaker {
     _posthook = reverting_posthook;
```
Except here -- we don't check the posthook was called

espendk replied:
Not sure how we can check it, since the posthook reverts and thus we cannot retain any state change made in it?

adhusson replied:
I tried to check Mangrove's corresponding log: 3994d96ff8b77f2486f322254aa3c1860b4ad9a5

adhusson commented on test/core/Scenarii.t.sol:
```diff
@@ -273,62 +269,63 @@ contract ScenariiTest is MangroveTest {
-    (MgvStructs.OfferUnpacked memory ofr,) = mgv.offerInfo($(base), $(quote), bag.snipedId);
-    assertTrue(ofr.gives == 0, "Offer should not have a residual");
-  }
+  // FIXME: Can this meaningfully be converted to clean and/or marketOrder?
```
For this I think we need the advice of @jkrivine

adhusson commented on test/core/TakerOperations.t.sol:
```diff
@@ -77,99 +80,32 @@ contract TakerOperationsTest is MangroveTest {
-    assertTrue((got == gave && gave == 0), "Taker should not give or take anything");
-  }
-
+  // FIXME: Not sure what this tests?
```
The purpose of this test is to make sure inbound volumes are rounded up when partially taking an offer i.e. you can't have the taker pay 0 if the maker sends > 0 to the taker.

The test sets this up with a wants=9, gives=10 offer,and the taker asks for a volume of 1.

espendk replied:
Ok, thanks for the explanation :-) I've copied it into a comment, see [8f2c2cb](https://github.com/mangrovedao/mangrove-core/pull/454/commits/8f2c2cb6413e55228132de90f317cde596c0fde1)

adhusson commented on test/core/TakerOperations.t.sol:
```diff
@@ -77,99 +80,32 @@ contract TakerOperationsTest is MangroveTest {
     assertEq(gave, 0.5 ether, "Taker did not give correct amount");
   }
 
-  function test_multiple_snipes_fillWants() public {
```
Is there a `clean` version of this test that should exist somewhere?

espendk replied:
Yes, `test_clean_multiple_failing_offers` in the same file.

adhusson commented on test/core/TakerOperations.t.sol:
```diff
@@ -197,54 +130,86 @@ contract TakerOperationsTest is MangroveTest {
     assertTrue(!mgv.offers($(base), $(quote), ofr).isLive(), "Offer should not be in the book");
   }
 
-  function test_snipe_fillGives_zero() public {
```
Similarly, shouldn't we test something similar with clean (ie that cleaning fails in this setup)?

espendk replied:
The test `test_cleaning_non_failing_offer_fails` tests the same, just in a slightly different way (it was written before migrating the snipes tests).

adhusson commented on test/core/TakerOperations.t.sol:
```diff
@@ -297,14 +280,15 @@ contract TakerOperationsTest is MangroveTest {
-    (uint successes, uint takerGot, uint takerGave,,) =
-      testMgv.snipesInTest($(base), $(quote), wrap_dynamic([ofr, uint(Tick.unwrap(offerTick)), 1 ether, 100_000]), true);
+    vm.expectEmit(true, false, false, false, $(mgv));
+    emit Credit($(refusemkr), 0); // credited amount should be "mkr_provision - penalty"; Not including this in the emit assertion, as penalty is difficult to maintain.
```
I don't understand this comment

espendk replied:
Ideally, we'd include the credited value in the emitted event. However, the penalty isn't something we can calculate (since it depends on actual gas usage) and I think it's likely to change over time, so putting in a constant would require maintenance.

Do you think we should something different?

espendk replied:
I found this pattern in `MakerPosthook.t.sol`, I'll copy that instead:

```
    //TODO: when events can be checked instead of expected, take given penalty instead of ignoring it
    vm.expectEmit(true, true, true, false, $(mgv));
```

espendk replied:
Addressed in [a21d89d](https://github.com/mangrovedao/mangrove-core/pull/454/commits/a21d89dbc0f556eda8cdea09e69e21f89da9b907)

adhusson commented on test/core/TakerOperations.t.sol:
```diff
@@ -330,18 +314,17 @@ contract TakerOperationsTest is MangroveTest {
-    (uint successes, uint takerGot, uint takerGave,,) =
-      testMgv.snipesInTest($(base), $(quote), wrap_dynamic([ofr, uint(Tick.unwrap(offerTick)), 1 ether, 100_000]), true);
+    vm.expectEmit(true, false, false, false, $(mgv));
+    emit Credit($(mkr), 0); // credited amount should be "mkr_provision - penalty"; Not including this in the emit assertion, as penalty is difficult to maintain.
```
Same as above!

espendk replied:
Same answer/question as above :-)

espendk replied:
Addressed in [a21d89d](https://github.com/mangrovedao/mangrove-core/pull/454/commits/a21d89dbc0f556eda8cdea09e69e21f89da9b907)

adhusson commented on test/core/TakerOperations.t.sol:
```diff
@@ -641,37 +593,15 @@ contract TakerOperationsTest is MangroveTest {
-    assertTrue(successes == 0, "Order should fail when order price is higher than offer");
-  }
-
-  function test_snipe_on_higher_gas_fails() public {
```
Do we have tests for clean on that topic?

espendk replied:
Yes, `test_giving_smaller_gasreq_to_clean_fails` in the same file.

adhusson replied:
Sorry I should have checked better that all those tests I asked about had cleaning counterparts!

espendk replied:
np :-)

espendk committed:
test: Assert that maker{Execute,Posthook} for offers are (not) called

After migrating away from `snipes` it was not clear that offers were executed as expected, leaving tests possibly vacuously satisfied. Relevant tests have therefore been augmented with assertions that maker{Execute,Posthook} is successfully executed.

To ease this, TestMaker has been extended with facilities for tracking successful execution of makerExecute and makerPosthook.

espendk committed:
doc(test): Clarify comments re. penalty in expected events

espendk committed:
doc(test): Explain the `test_taker_cannot_drain_maker` test

adhusson committed:
test: test posthook failure is logged by mgv

adhusson reviewed (approved):
pending my small test-editing commit lgtm

espendk commented:
> pending my small test-editing commit lgtm

Perfect, thanks.

I'll merge once @jkrivine has had a chance to weigh in re. Scenarii.t.sol.

espendk commented:
I merged the PR with the FIXME re. Scenarii in as @jkrivine  is busy and we need to move on :-)
We'll address the Scenarii test later - I'll add a task on it to the ticks PR.
